LuaRules/Gadgets/customCommands.lua

```diff
@@ -0,0 +1,55 @@
+local moduleInfo = {
+	name = "customCommandsGadget",
+	desc = "Gadget connecting custom commands module allowing players to define their own commands",
+	author = "PepeAmpere",
+	date = "2017-01-24",
+	license = "notAlicense",
+	layer = 0,
+	enabled = true -- loaded by default?
+}
+
+function gadget:GetInfo()
+	return moduleInfo
+end
+
+-- get madatory module operators
+VFS.Include("modules.lua") -- modules table
+VFS.Include(modules.attach.data.path .. modules.attach.data.head) -- attach lib module
+
+-- get other madatory dependencies
+attach.Module(modules, "message")
+attach.Module(modules, "customCommands")
+
+-- UNSYNCED
+if (not gadgetHandler:IsSyncedCode()) then
+	function gadget:Initialize()
+		gadgetHandler:AddSyncAction('CustomCommandUpdate', receiveCustomMessage.CustomCommandUpdate)
+		gadgetHandler:AddSyncAction('CustomCommandRegistered', receiveCustomMessage.CustomCommandRegistered)
+	end
+	
+	return
+end
+
+-- SYNCED
+
+-- units incomming
+function gadget:UnitCreated(unitID, unitDefID, unitTeam, builderID)
+	customCommands.UnitIncomming(unitID, unitDefID, unitTeam)
+end
+function gadget:UnitGiven(unitID, unitDefID, newTeam, oldTeam) -- in this moment the unit is already part of the newTeam
+	customCommands.UnitIncomming(unitID, unitDefID, newTeam)
+end
+
+-- units removed
+function gadget:UnitDestroyed(unitID, unitDefID, unitTeam, attackerID, attackerDefID, attackerTeamD)
+	customCommands.UnitRemoved(unitID, unitDefID, unitTeam)
+end
+function gadget:UnitTaken(unitID, unitDefID, oldTeam, newTeam) -- in this moment the unit is still part of the oldTeam
+	customCommands.UnitRemoved(unitID, unitDefID, oldTeam)
+end
+
+-- msg updates
+function gadget:RecvLuaMsg(msg, playerID)
+	message.Receive(msg, playerID) -- using messageReceiver data structure
```

Comment:
Nice topic, thanks for linking it.

I'm aware of serialization not being much effective from performance/memory consumption perspective. Anywhere it is used, it is used for reasonable tasks :) = we send tables of low complexity never transferring e.g. huge objects (e.g. chili objects with huge metatables). In all cases its add hoc constructed table, usually 1-2 levels deep. 

Even the support for Vec3 and HMSF (time vector) I added into message system (and its [serialization and deserialization logic](https://bitbucket.org/notabots/core/commits/54457cd75e1ee45283b4ae117c92cf518b8e4216)) lately wont transfer any methods or metatables in message. E.g. Vector is translated into simple table with special flag telling the deserialization logic to create vector instead of simple table on the other side of the tunnel.

The key is to transfer the bare minimum. If sender and receiver both know the vector, i dont need to send the object properties via the tunnel ;).